keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2262 @@
+	// a user might have selected a node manually without reserving
+	// resources or a
+	// scheduler might have missed that step, for example because it
+	// doesn't support dynamic resource allocation or the feature was
+	// disabled.
+	ReservedFor []metav1.OwnerReference
+}
+
+// SchedulingStatus contains information that is relevant while
+// a Pod with delayed allocation is being scheduled.
+type SchedulingStatus struct {
+	// When allocation is delayed, the scheduler must set
+	// the node for which it wants the resource to be allocated
+	// before the driver proceeds with allocation.
+	//
+	// This field may only be set by a scheduler, but not get
+	// overwritten. It will get reset by the driver when allocation
+	// succeeds or fails. This ensures that different schedulers
+	// that handle different Pods do not accidentally trigger
+	// allocation for different nodes.
+	//
+	// For immediate allocation, the scheduler will not set
+	// this field. The resource driver controller may
+	// set it to trigger allocation on a specific node if the
+	// resources are local to nodes.
+	//
+	// List/watch requests for ResourceClaims can filter on this field
+	// using a "status.scheduling.scheduler.selectedNode=NAME"
+	// fieldSelector.
+	SelectedNode string
+
+	// SelectedUser may be set by the scheduler together with SelectedNode
+	// to the Pod that it is scheduling. The resource driver then may set
+	// both the Allocation and the ReservedFor field when it is done with
+	// a successful allocation.
+	//
+	// This is an optional optimization that saves one API server call
+	// and one pod scheduling attempt in the scheduler because the resource
+	// will be ready for use by the Pod the next time that the scheduler
+	// tries to schedule it.
+	SelectedUser *metav1.OwnerReference
+}
+
+// AllocationResult contains attributed of an allocated resource.
+type AllocationResult struct {
+	// Attributes contains arbitrary data returned by the driver after a
+	// successful allocation.  This data is passed to the driver for all
+	// operations involving the allocated resource. This is opaque for
+	// Kubernetes.  Driver documentation may explain to users how to
+	// interpret this data if needed.
+	//
+	// The attributes must be sufficient to deallocate the resource because
+	// the ResourceClass might not be available anymore when deallocation
+	// starts.
+	Attributes map[string]string
+
+	// This field will get set by the resource driver after it has
+	// allocated the resource driver to inform the scheduler where it can
+	// schedule Pods using the ResourceClaim.
+	//
+	// A resource driver may already set this before the resource is
+	// allocated. The scheduler will then check this field in addition to
+	// UnsuitableNodes to filter out nodes where the resource cannot be
+	// allocated.
+	//
+	// Setting this field is optional. If nil, the resource is available
+	// everywhere.
+	AvailableOnNodes *core.NodeSelector
```

Comment:
This is the part that worries me most, actually. Plugins for devices will want to start setting labels on nodes, and there's no discussion of admission control or access control to limit what label keys a plugin could write to.

We could end up with: a lot of cardinality; device plugins fighting to control private labels; too-wide RBAC rules.

My worry is that this design tends to encourage what I might term misbehavior. Is there a different approach that puts the cluster admin more obviously in charge of deciding what labels each plugin gets to assert?

Reply:
To be clear: I'm not worried about _well behaved_ drivers / extensions. It's definitely the other kind that concern me.

Reply:
I don't see how this KEP could be changed to prevent writing bad drivers. Removing NodeSelector and using something else will not change that a resource driver needs to be installed according to the install instructions of the vendor and then may be able to all kinds off bad things.

Reply:
I agree - a driver that wants to use node labels will have to ask the admin for node RBAC or admin will do it manually.  That seems acceptable to me.

Reply:
Fair enough, I'm happy to have this one marked resolved then.

(not sure about the field name either, but we can revise that before GA without breaking clients).